jlm/llvm/ir/operators/MemCpy.hpp

```diff
@@ -0,0 +1,188 @@
+    auto resultPorts = CreateResultPorts(memoryStates.size());
+
+    std::vector<const variable *> operands = { destination, source, length };
+    operands.insert(operands.end(), memoryStates.begin(), memoryStates.end());
+
+    MemCpyOperation op(operandPorts, resultPorts);
+    return tac::create(op, operands);
+  }
+
+  static std::vector<rvsdg::output *>
+  create(
+      rvsdg::output * destination,
+      rvsdg::output * source,
+      rvsdg::output * length,
+      const std::vector<rvsdg::output *> & memoryStates)
+  {
+    auto operandPorts = CheckAndCreateOperandPorts(length->type(), memoryStates.size());
+    auto resultPorts = CreateResultPorts(memoryStates.size());
+
+    std::vector<rvsdg::output *> operands = { destination, source, length };
+    operands.insert(operands.end(), memoryStates.begin(), memoryStates.end());
+
+    MemCpyOperation op(operandPorts, resultPorts);
+    return rvsdg::simple_node::create_normalized(destination->region(), op, operands);
+  }
+
+private:
+  static std::vector<rvsdg::port>
+  CheckAndCreateOperandPorts(const rvsdg::type & length, size_t numMemoryStates)
+  {
+    if (length != rvsdg::bit32 && length != rvsdg::bit64)
+      throw util::error("Expected 32 bit or 64 bit integer type.");
+
+    if (numMemoryStates == 0)
+      throw util::error("Number of memory states cannot be zero.");
+
+    PointerType pointerType;
+    std::vector<rvsdg::port> ports = { pointerType, pointerType, length };
+    ports.insert(ports.end(), numMemoryStates, { MemoryStateType::Create() });
+
+    return ports;
+  }
+
+  static std::vector<rvsdg::port>
+  CreateResultPorts(size_t nMemoryStates)
+  {
+    return std::vector<rvsdg::port>(nMemoryStates, { MemoryStateType::Create() });
+  }
+};
+
+/**
+ * Represents a volatile LLVM memcpy intrinsic
+ *
+ * In contrast to LLVM, a volatile memcpy requires in an RVSDG setting an I/O state as it
+ * incorporates externally visible side-effects. This I/O state allows the volatile memcpy operation
+ * to be sequentialized with respect to other volatile memory accesses and I/O operations. This
+ * additional I/O state is the main reason why volatile memcpys are modeled as its own operation and
+ * volatile is not just a flag at the normal \ref MemCpyOperation.
+ *
+ * @see MemCpyOperation
+ */
+class MemCpyVolatileOperation final : public rvsdg::simple_op
+{
+public:
+  ~MemCpyVolatileOperation() noexcept override;
+
+  MemCpyVolatileOperation(const rvsdg::type & lengthType, size_t numMemoryStates)
+      : rvsdg::simple_op(
+          CheckAndCreateOperandPorts(lengthType, numMemoryStates),
+          CreateResultPorts(numMemoryStates))
+  {}
+
+  bool
+  operator==(const operation & other) const noexcept override;
+
+  [[nodiscard]] std::string
+  debug_string() const override;
+
+  [[nodiscard]] std::unique_ptr<rvsdg::operation>
+  copy() const override;
+
+  static std::unique_ptr<llvm::tac>
+  CreateThreeAddressCode(
```

Comment:
I assume there will be a future cleanup changing the names of the corresponding methods in `MemCpyOperation`. You have probably already confirmed this when I've asked about it previously.

Reply:
A future cleanup will collect all the creator methods for three address codes and RVSDG nodes in a builder, respectively. This will delegate the responsible to a single entity instead of spreading it out all over as well as save some code (as checks and port creation etc. can be unified for all operations instead of each of them implementing their own).